### PR TITLE
Feature/sa 27 scheduler

### PR DESCRIPTION
**Pull request by progers-tech**

### CODE REVIEW

File: infrastructure/infrastructure.tf [hunk 1]
```diff
@@ -89,3 +94,40 @@ module "nmap_task" {
...
+
+  app_name            = "${var.app_name}"
+  task_name           = "${var.task_name}"
+  scan_hosts          = "${var.scan_hosts}"
+  queue_url           = "${module.nmap_task.task_queue_url}"
+  queue_arn           = "${module.nmap_task.task_queue}"
+  transient_workspace = "${!contains(var.known_deployment_stages, terraform.workspace)}"
+}
+
+data "aws_iam_policy_document" "nmap_cwe_sqs_policy" {
```
Comment (blackmamo) [inline infrastructure/infrastructure.tf]: Since this has no other dependencies, it should be added to the task execution module, not the nmap

File: infrastructure/infrastructure.tf [hunk 2]
```diff
@@ -89,3 +94,40 @@ module "nmap_task" {
...
+    principals {
+      type        = "Service"
+      identifiers = ["events.amazonaws.com"]
+    }
+
+    resources = ["${module.nmap_task.task_queue}"]
+  }
+}
+
+resource "aws_sqs_queue_policy" "nmap_cwe_sqs_rule" {
```
Comment (blackmamo) [inline infrastructure/infrastructure.tf]: This will go away too

File: infrastructure/infrastructure.tf (hunk 1 above)
Comment (blackmamo) [reply infrastructure/infrastructure.tf]: Add this to the existing notify_topic_policy

File: serverless.yml [hunk 3]
```diff
@@ -78,7 +78,7 @@ provider:
          Fn::Join:
            - ""
            - - ${ssm:/${self:custom.appName}/${self:custom.deployStage}/tasks/${self:custom.pluginName}/s3/results/arn}
-
+    
 plugins:
```
Comment (blackmamo) [inline serverless.yml]: Non semantic change, please remove

File: tests/task_queue_consumer/task_queue_consumer_test.py [hunk 4]
```diff
@@ -28,11 +28,14 @@ def ssm_return_vals(using_private):
     ssm_prefix = f"/{app_name}/{stage}"
     return {
         'Parameters': [
-            {"Name": f"{ssm_prefix}/vpc/using_private_subnets", "Value": "true" if using_private else "false"},
-            {"Name": f"{ssm_prefix}/tasks/{task_name}/security_group/id", "Value": "sg-123"},
+            {"Name": f"{ssm_prefix}/vpc/using_private_subnets",
```
Comment (blackmamo) [inline tests/task_queue_consumer/task_queue_consumer_test.py]: Formatting only changes adding noise, what is your line length setup? Can we avoid these please?

File: tests/task_queue_consumer/task_queue_consumer_test.py [hunk 5]
```diff
@@ -66,9 +69,10 @@ def expected_params(public_ip_str, scan_targets, message_id):
 
 @pytest.mark.unit
 @serialise_mocks()
-@resetting_mocks(task_queue_consumer.ecs_client, task_queue_consumer.ssm_client)
+@resetting_mocks(task_queue_consumer.ecs_client, task_queue_consumer.ssm_client.get_parameters)
```
Comment (blackmamo) [inline tests/task_queue_consumer/task_queue_consumer_test.py:69]: We said that you need to change these to:

@resetting_mocks(task_queue_consumer.ecs_client.run_task, task_queue_consumer.ssm_client.get_parameters)

File: tests/task_queue_consumer/task_queue_consumer_test.py [hunk 6]
```diff
@@ -78,9 +82,10 @@ def test_task_queue_consumer_private_subnet():
...
```
Comment (blackmamo) [inline tests/task_queue_consumer/task_queue_consumer_test.py:81]: same as above and in other places below

File: tests/task_queue_consumer/task_queue_consumer_test.py [hunk 7]
```diff
@@ -78,9 +82,10 @@ def test_task_queue_consumer_private_subnet():
...
 def test_task_queue_consumer_no_subnet():
-    task_queue_consumer.ssm_client.get_parameters.return_value = ssm_return_vals(False)
+    task_queue_consumer.ssm_client.get_parameters.return_value = ssm_return_vals(
```
Comment (blackmamo) [inline tests/task_queue_consumer/task_queue_consumer_test.py]: Line length changes here too. Lets please agree on a line length and avoid these

File: tests/task_queue_consumer/task_queue_consumer_test.py [hunk 8]
```diff
@@ -188,22 +205,105 @@ def test_sanitises_input_dodgy_input():
...
     with pytest.raises(RuntimeError, match=r"\{\"arn\": \"arn::some:::arn\", \"reason\": \"failed miserably\"\}"):
-        task_queue_consumer.submit_scan_task({"Records": [{"body": "some.host", "messageId": "13"}]}, mock.MagicMock())
+        task_queue_consumer.submit_scan_task(
+            {"Records": [{"body": "some.host", "messageId": "13"}]}, mock.MagicMock())
+
+
+@pytest.mark.unit
+@serialise_mocks()
+@resetting_mocks(task_queue_consumer.ecs_client, task_queue_consumer.ssm_client.get_parameters)
+def test_sanitises_input_ip4_json():
```
Comment (blackmamo) [inline tests/task_queue_consumer/task_queue_consumer_test.py]: CAn we change the names of these 4 tests please? They aren't really testing the sanitising, they are testing alternative input syntax and grouping

**Review by blackmamo — CHANGES_REQUESTED** [review]

File: infrastructure/infrastructure.tf (hunk 1 above)
Comment (progers-tech) [reply infrastructure/infrastructure.tf]: moved to trigger_queue in task execution

File: infrastructure/infrastructure.tf (hunk 2 above)
Comment (progers-tech) [reply infrastructure/infrastructure.tf]: this too

File: task_queue_consumer/task_queue_consumer.py [hunk 9]
```diff
@@ -81,49 +130,22 @@ def sanitise_nmap_target(target_str):
...
-            raise RuntimeError(f"ECS task failed to start {dumps(failures)}")
+        if record['body'][0] == '{':
+            # triggered from cloudwatch which requires format in JSON:
+            body = loads(record['body'])
+            print(body)
+            if 'CloudWatchEventHosts' in body:
+                id = 0
+                for host in body['CloudWatchEventHosts']:
+                    id += 1
+                    print(host)
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: CAn we not just print the host nakedly? Either remove the line or add some context please

File: task_queue_consumer/task_queue_consumer.py [hunk 10]
```diff
@@ -81,49 +130,22 @@ def sanitise_nmap_target(target_str):
...
-        failures = task_response["failures"]
-        if len(failures) != 0:
-            raise RuntimeError(f"ECS task failed to start {dumps(failures)}")
+        if record['body'][0] == '{':
+            # triggered from cloudwatch which requires format in JSON:
+            body = loads(record['body'])
+            print(body)
+            if 'CloudWatchEventHosts' in body:
+                id = 0
+                for host in body['CloudWatchEventHosts']:
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: host is probably not a good name since there could still be multiple space separated hosts and networks. Call it nmap_targets, or nmap_target_strings, or something?

File: task_queue_consumer/task_queue_consumer.py [hunk 11]
```diff
@@ -65,11 +66,59 @@ def sanitise_nmap_target(target_str):
...
+        "launchType": "FARGATE",
+        "overrides": {
+            "containerOverrides": [{
+                "name": task_name,
+                "environment": [
+                    # TODO The only bit of this file that isn't going to be the same for other
+                    # task queue executors, is this bit that maps the request body to some env vars
+                    # Extract the common code into a layer exported by the task-execution project
+                    {
+                        "name": "HOST_TO_SCAN",
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: Should probably change this name to NMAP_TARGET_STRING too, based on my other comment. Will need to match that in the task script too.

File: task_queue_consumer/task_queue_consumer.py [hunk 12]
```diff
@@ -65,11 +66,59 @@ def sanitise_nmap_target(target_str):
...
+                                "value": host
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: Why did you move the sanitise from here to the submit_scan_task method? You have to call it in two separate places now...

File: task_queue_consumer/task_queue_consumer.py [hunk 13]
```diff
@@ -81,49 +130,22 @@ def sanitise_nmap_target(target_str):
...
+        if record['body'][0] == '{':
+            # triggered from cloudwatch which requires format in JSON:
+            body = loads(record['body'])
+            print(body)
+            if 'CloudWatchEventHosts' in body:
+                id = 0
+                for host in body['CloudWatchEventHosts']:
+                    id += 1
+                    print(host)
+                    submit_ecs_task(event,  sanitise_nmap_target(
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: push the sanitise back into the submit_ecs_task method

File: task_queue_consumer/task_queue_consumer.py [hunk 14]
```diff
@@ -81,49 +130,22 @@ def sanitise_nmap_target(target_str):
...
+                        host.strip()), f'{record["messageId"]}-{id}')
+        else:
+            # triggered via AWS by pushing host(s) to the queue manually:
+            submit_ecs_task(event, sanitise_nmap_target(
```
Comment (blackmamo) [inline task_queue_consumer/task_queue_consumer.py]: push the sanitise back into the submit_ecs_task method

**Review by blackmamo — CHANGES_REQUESTED** [review]

File: task_queue_consumer/task_queue_consumer.py (hunk 9 above)
Comment (progers-tech) [reply task_queue_consumer/task_queue_consumer.py]: deleted

File: .circleci/config.yml [hunk 15]
```diff
@@ -54,6 +54,7 @@ jobs:
           command: |
             . .circleci/env.sh
             cd infrastructure
+            echo $CIRCLE_HOSTS >scan_hosts.auto.tfvars
```
Comment (blackmamo) [inline .circleci/config.yml]: minor - space please

File: infrastructure/docker_image/Dockerfile.template [hunk 16]
```diff
@@ -8,7 +8,7 @@ RUN mv vulners.nse /usr/share/nmap/scripts
 
 COPY task_script.sh .
 
-ENV HOST_TO_SCAN=127.0.0.1
+ENV NMAP_TARGET_STRING=127.0.0.1
```
Comment (blackmamo) [inline infrastructure/docker_image/Dockerfile.template:11]: 👍

File: infrastructure/infrastructure.tf [hunk 17]
```diff
@@ -75,7 +80,7 @@ module "elastic_resources" {
 }
 
 module "nmap_task" {
-  source = "github.com/ministryofjustice/securityanalytics-taskexecution//infrastructure/ecs_task"
```
Comment (blackmamo) [inline infrastructure/infrastructure.tf]: Do not do this!!!

File: infrastructure/infrastructure.tf [hunk 18]
```diff
@@ -97,3 +102,21 @@ module "nmap_task" {
   ssm_source_stage              = "${local.ssm_source_stage}"
   transient_workspace           = "${!contains(var.known_deployment_stages, terraform.workspace)}"
 }
+
+module "nmap_task_scheduler" {
+  source = "github.com/ministryofjustice/securityanalytics-taskexecution/infrastructure/scheduler"
```
Comment (blackmamo) [inline infrastructure/infrastructure.tf]: Add double slash here

File: infrastructure/infrastructure.tf (hunk 17 above)
Comment (blackmamo) [reply infrastructure/infrastructure.tf]: https://www.terraform.io/docs/modules/sources.html#modules-in-package-sub-directories

In fact add a comment so no one else undoes this

**Review by blackmamo — CHANGES_REQUESTED** [review]
Minor only

**Review by blackmamo — APPROVED** [review]
